source/PluginDev/Assets/GooglePlayGames/Platforms/Android/AndroidRealTimeMultiplayerClient.cs

```diff
@@ -34,111 +54,719 @@ public AndroidRealTimeMultiplayerClient(AndroidJavaObject account)
                     roomConfigClass.CallStatic<AndroidJavaObject>("builder", roomUpdateCallback))
                 {
-                    roomConfigBuilder.Call<AndroidJavaObject>("setVariant", (int) variant);
-                    roomConfigBuilder.Call<AndroidJavaObject>("setAutoMatchCriteria",
-                        roomConfigBuilder.CallStatic<AndroidJavaObject>("createAutoMatchCriteria", (int) minOpponents,
-                            (int) maxOpponents, (long) exclusiveBitMask));
+                    if (variant > 0)
+                    {
+                        roomConfigBuilder.Call<AndroidJavaObject>("setVariant", (int) variant);
+                    }
+
+                    using (var autoMatchCriteria = roomConfigClass.CallStatic<AndroidJavaObject>(
+                        "createAutoMatchCriteria", (int) minOpponents,
+                        (int) maxOpponents, (long) exclusiveBitMask))
+                    using (roomConfigBuilder.Call<AndroidJavaObject>("setAutoMatchCriteria", autoMatchCriteria))
+                        ;
+
+                    using (var messageReceivedListener = new AndroidJavaObject(
+                        "com.google.games.bridge.RealTimeMessageReceivedListenerProxy",
+                        new MessageReceivedListenerProxy(listener)))
+                    using (roomConfigBuilder.Call<AndroidJavaObject>("setOnMessageReceivedListener",
+                        messageReceivedListener))
+                        ;
+
+                    using (var roomStatusUpdateCallback = new AndroidJavaObject(
+                        "com.google.games.bridge.RoomStatusUpdateCallbackProxy",
+                        new RoomStatusUpdateCallbackProxy(this, listener)))
+                    using (roomConfigBuilder.Call<AndroidJavaObject>("setRoomStatusUpdateCallback",
+                        roomStatusUpdateCallback))
+                        ;
+
                     mRoomConfig = roomConfigBuilder.Call<AndroidJavaObject>("build");
+                    mListener = listener;
+                }
+
+                using (var task = mClient.Call<AndroidJavaObject>("create", mRoomConfig))
+                {
+                    AndroidTaskUtils.AddOnFailureListener(
+                        task,
+                        e =>
+                        {
+                            listener.OnRoomConnected(false);
+                            CleanSession();
+                        });
+                }
+            }
+        }
+
+        public void CreateWithInvitationScreen(uint minOpponents, uint maxOpponents, uint variant,
+            RealTimeMultiplayerListener listener)
+        {
+            AndroidHelperFragment.ShowSelectOpponentsUI(minOpponents, maxOpponents, /* isRealTime= */ true,
+                (status, result) =>
+                {
+                    if (status != UIStatus.Valid)
+                    {
+                        listener.OnRoomConnected(false);
+                        CleanSession();
```

Comment:
Should this CleanSession be in lock section?

Reply:
Agree, CleanSession should be in lock section. But instead of adding it here, I added lock section inside CleanSession.